### PR TITLE
feat: add test proxy to java-bigtable client

### PR DESCRIPTION
**Pull request by mutianf**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: test-proxy/pom.xml [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.google.cloud</groupId>
+  <artifactId>google-cloud-bigtable-test-proxy</artifactId>
+  <version>1.0</version>
+  <packaging>jar</packaging>
+  <name>Google Cloud Bigtable Test Proxy</name>
+  <url>https://github.com/googleapis/java-bigtable</url>
+
```
Comment (igorbernstein2) [inline test-proxy/pom.xml:10]: This definitely needs a description

File: test-proxy/pom.xml [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        <artifactId>google-cloud-bigtable-deps-bom</artifactId>
+        <version>2.15.1-SNAPSHOT</version><!-- {x-version-update:google-cloud-bigtable:current} -->
+        <type>pom</type>
+        <scope>import</scope>
+      </dependency>
+      <dependency>
+        <groupId>com.google.auto.value</groupId>
+        <artifactId>auto-value</artifactId>
+        <version>1.9</version>
+      </dependency>
```
Comment (igorbernstein2) [inline test-proxy/pom.xml:37]: Here and everywhere else autovalue is mentioned in the pom:

I think you want to remove this and just add a ENableAutoValue file:

https://github.com/googleapis/java-shared-config/blob/main/pom.xml#L734-L779

File: test-proxy/pom.xml [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        <configuration>
+          <protocArtifact>com.google.protobuf:protoc:3.9.0:exe:${os.detected.classifier}</protocArtifact>
+          <pluginId>grpc-java</pluginId>
+          <pluginArtifact>io.grpc:protoc-gen-grpc-java:1.24.0:exe:${os.detected.classifier}</pluginArtifact>
+        </configuration>
+        <executions>
+          <execution>
+            <goals>
+              <goal>compile</goal>
+              <goal>compile-custom</goal>
```
Comment (igorbernstein2) [inline test-proxy/pom.xml:86]: whats compile-custom?

File: test-proxy/pom.xml [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
+        <version>3.2.4</version>
+        <executions>
+          <execution>
+            <phase>package</phase>
```
Comment (igorbernstein2) [inline test-proxy/pom.xml:97]: why does this need to be shaded?

File: test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+      responseObserver.onError(
+          Status.ALREADY_EXISTS
+              .withDescription("Client " + request.getClientId() + " already exists.")
+              .asException());
+      return;
+    }
+
+    BigtableDataSettings.Builder settingsBuilder = BigtableDataSettings.newBuilder();
+    if (request.hasPerOperationTimeout()) {
+      com.google.protobuf.Duration timeoutFromReq = request.getPerOperationTimeout();
```
Comment (igorbernstein2) [inline test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java:225]: Can you simplify this to Duration newTimeout = Duration.fromMillis(Durations.toMillis(request.getPerOperationTimeout());

File: test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java [hunk 6]
```diff
@@ -0,0 +1,768 @@
...
+    }
+
+    responseObserver.onNext(CreateClientResponse.getDefaultInstance());
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void closeClient(
+      CloseClientRequest request, StreamObserver<CloseClientResponse> responseObserver) {
+    CbtClient client = idClientMap.get(request.getClientId());
```
Comment (igorbernstein2) [inline test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java:260]: Why not just call getClient()?

File: test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+  public void mutateRow(
+      MutateRowRequest request, StreamObserver<MutateRowResult> responseObserver) {
+    CbtClient client = idClientMap.get(request.getClientId());
+    if (client == null) {
+      responseObserver.onError(
+          Status.NOT_FOUND
+              .withDescription("Client " + request.getClientId() + " not found.")
+              .asException());
+      return;
+    }
```
Comment (igorbernstein2) [inline test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java:301]: here and everywhere else...just why not just use getCLient?

File: test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java [hunk 8]
```diff
@@ -0,0 +1,768 @@
...
+  }
+
+  private CredentialsProvider getCredentialsProvider() throws IOException {
+    if (credential == null) {
+      return NoCredentialsProvider.create();
+    }
+
+    final ServiceAccountJwtAccessCredentials creds =
+        ServiceAccountJwtAccessCredentials.fromStream(
+            new ByteArrayInputStream(credential.getBytes(UTF_8)));
```
Comment (igorbernstein2) [inline test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxy.java:754]: why are you hardcoding jwt creds? I think GoogleCredentials.fromStream() is more appropriate

File: test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxyMain.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+  private static final Logger logger = Logger.getLogger(CbtTestProxyMain.class.getName());
+
+  public static void main(String[] args) throws InterruptedException, IOException {
+    int port = Integer.parseInt(System.getProperty("port", "9999"));
+    if (port <= 0) {
+      throw new IllegalArgumentException(String.format("Port %d is not > 0.", port));
+    }
+
+    CbtTestProxy cbtTestProxy = CbtTestProxy.createUnencrypted();
```
Comment (igorbernstein2) [inline test-proxy/src/main/java/com/google/cloud/bigtable/testproxy/CbtTestProxyMain.java:35]: should this be in an else?

File: test-proxy/pom.xml (hunk 1 above)
Comment (kolea2) [reply test-proxy/pom.xml:10]: @igorbernstein2 @mutianf are we planning on publishing this module?

Comment (mutianf) [reply test-proxy/pom.xml:10]: I think so, is there any concern with adding this in public repo?

File: test-proxy/pom.xml (hunk 3 above)
Comment (mutianf) [reply test-proxy/pom.xml:86]: It's from the example in grpc-java: https://github.com/grpc/grpc-java/blob/v1.50.2/examples/pom.xml#L106

File: test-proxy/pom.xml (hunk 4 above)
Comment (mutianf) [reply test-proxy/pom.xml:97]: To specify the main class to start the proxy I think?

File: test-proxy/pom.xml [hunk 10]
```diff
@@ -0,0 +1,126 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.google.cloud</groupId>
+  <artifactId>google-cloud-bigtable-test-proxy</artifactId>
+  <version>1.0</version>
```
Comment (kolea2) [inline test-proxy/pom.xml:6]: I would recommend changing this to 0.0.1-SNAPSHOT, and this also needs to be added to versions.txt

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: test-proxy/pom.xml (hunk 10 above)
Comment (mutianf) [reply test-proxy/pom.xml:6]: I don't see samples in versions.txt though?

File: test-proxy/pom.xml (hunk 1 above)
Comment (igorbernstein2) [reply test-proxy/pom.xml:11]: I dont see any reason not to

File: test-proxy/pom.xml (hunk 3 above)
Comment (igorbernstein2) [reply test-proxy/pom.xml:78]: I see compile is for protobufs and custom is for grpc stubs

File: test-proxy/pom.xml (hunk 4 above)
Comment (igorbernstein2) [reply test-proxy/pom.xml:88]: main entry point in the jar can be specified to make it executable:
https://maven.apache.org/shared/maven-archiver/examples/classpath.html#Make

Shading will produce a single jar thats executable. There is no readme here so I'm not sure how this is going to be executed. So I'll list a couple of approaches:

1. Having a bash script that will mvn exec:java the proxy and then start the test suite
2. Have a junit wrapper that starts the proxy at the beginning of the test run and then exec's the test suite

In both cases, we don't need to bundle the dependencies. In fact, bundling the deps prevents us from running the proxy with older clients

File: test-proxy/pom.xml [hunk 11]
```diff
@@ -3,10 +3,11 @@
...
   <artifactId>google-cloud-bigtable-test-proxy</artifactId>
-  <version>1.0</version>
+  <version>0.0.1-SNAPSHOT</version>
   <packaging>jar</packaging>
   <name>Google Cloud Bigtable Test Proxy</name>
   <url>https://github.com/googleapis/java-bigtable</url>
+  <description>Cloud Bigtable Java Client test proxy for running conformance tests.</description>
 
   <parent>
     <artifactId>google-cloud-bigtable-parent</artifactId>
```
Comment (liujiongxin) [inline test-proxy/pom.xml:13]: Does the version tag below accept custom values (i.e., we want to test a version released a while ago)?

**Review by liujiongxin — COMMENTED** [review]
Overall look good to me! Thanks for the effort!

File: test-proxy/pom.xml (hunk 1 above)
Comment (mutianf) [reply test-proxy/pom.xml:11]: To maven central? I told Kristen that we're not going to 😅

Comment (mutianf) [reply test-proxy/pom.xml:11]: Chatted offline, decided to not publishing it for now.

**Review by liujiongxin — COMMENTED** [review]
A readme file will be very helpful.

Example: https://github.com/googleapis/google-cloud-go/tree/main/bigtable/internal/testproxy, https://github.com/dbolduc/google-cloud-cpp/tree/cbt-test-proxy-dev-flattened/google/cloud/bigtable/cbt_test_proxy

File: test-proxy/pom.xml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+              <goal>compile-custom</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+      <!-- start skip publishing to maven central -->
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-deploy-plugin</artifactId>
+        <version>3.0.0-M2</version>
```
Comment (mutianf) [inline test-proxy/pom.xml:91]: we probably don't need this version

File: test-proxy/pom.xml [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+  <version>0.0.1-SNAPSHOT</version>
+  <packaging>jar</packaging>
+  <name>Google Cloud Bigtable Test Proxy</name>
+  <url>https://github.com/googleapis/java-bigtable</url>
+  <description>Cloud Bigtable Java Client test proxy for running conformance tests.</description>
+
+  <parent>
+    <artifactId>google-cloud-bigtable-parent</artifactId>
+    <groupId>com.google.cloud</groupId>
+    <version>2.15.1-SNAPSHOT</version><!-- {x-version-update:google-cloud-bigtable:current} -->
```
Comment (mutianf) [inline test-proxy/pom.xml:15]: update the parent version since we cut a new release

**Review by igorbernstein2 — APPROVED** [review]

File: test-proxy/README.md [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+Start the proxy on a different port
+
+```
+mvn exec:java -Dexec.mainClass=com.google.cloud.bigtable.testproxy.CbtTestProxyMain -Dport=1
+```
+
+Build and start the proxy with an older version of the client
+
+```
+mvn clean install -Dbigtable.client.version=2.1.0 -Denforcer.skip
```
Comment (kolea2) [inline test-proxy/README.md:37]: nit - I would replace 2.1.0 with <client_version> so the developer is intentional about which version they pick

**Review by kolea2 — APPROVED** [review]
lgtm with one small comment

**Pull request merged**